program-test/src/programs.rs

```diff
@@ -21,76 +26,198 @@ mod spl_associated_token_account {
+/// convenient to retrieve the ELF bytes directly.
+pub fn spl_program(
+    rent: &Rent,
+    target_id: Pubkey,
+) -> Result<(Vec<u8>, ProgramAccounts), ProgramLoadError> {
+    let target = SPL_PROGRAMS
+        .iter()
+        .find(|ProgramInfo { program_id, .. }| *program_id == target_id);
+
+    let target = target.ok_or(ProgramLoadError::UnknownId(target_id))?;
+
+    Ok((target.elf.to_vec(), program_accounts(rent, target)))
+}
+
+/// Returns accounts for all known programs, paired with their addresses.
+///
+/// Upgradable programs will have two accounts: one for the program and one for the program data.
+/// For upgradable programs the `Pubkey` returned is the public key of the main program account -
+/// just a you would normally expect.
 pub fn spl_programs(rent: &Rent) -> Vec<(Pubkey, AccountSharedData)> {
     SPL_PROGRAMS
         .iter()
-        .flat_map(|(program_id, loader_id, elf)| {
-            let mut accounts = vec![];
-            let data = if *loader_id == solana_sdk::bpf_loader_upgradeable::ID {
-                let (programdata_address, _) =
-                    Pubkey::find_program_address(&[program_id.as_ref()], loader_id);
-                let mut program_data = bincode::serialize(&UpgradeableLoaderState::ProgramData {
-                    slot: 0,
-                    upgrade_authority_address: Some(Pubkey::default()),
-                })
-                .unwrap();
-                program_data.extend_from_slice(elf);
-                accounts.push((
-                    programdata_address,
-                    AccountSharedData::from(Account {
-                        lamports: rent.minimum_balance(program_data.len()).max(1),
-                        data: program_data,
-                        owner: *loader_id,
-                        executable: false,
-                        rent_epoch: 0,
-                    }),
-                ));
-                bincode::serialize(&UpgradeableLoaderState::Program {
-                    programdata_address,
-                })
-                .unwrap()
-            } else {
-                elf.to_vec()
-            };
-            accounts.push((
-                *program_id,
-                AccountSharedData::from(Account {
-                    lamports: rent.minimum_balance(data.len()).max(1),
-                    data,
-                    owner: *loader_id,
-                    executable: true,
-                    rent_epoch: 0,
-                }),
-            ));
-            accounts.into_iter()
+        .flat_map(|info| {
+            let accounts = program_accounts(rent, info);
+            Vec::<_>::from(accounts).into_iter()
         })
         .collect()
 }
+
+fn program_accounts(rent: &Rent, info: &ProgramInfo) -> ProgramAccounts {
+    let ProgramInfo {
+        program_id,
+        loader_id,
+        elf,
+    } = info;
+
+    match *loader_id {
+        bpf_loader::ID => {
```

Comment:
The previous implementation was keying on only the upgradeable loader ID. I'm wondering if we should do the same thing here. It might make things easier now that BPF Loader is deprecated and Loader v4 is being rolled out soon. Also it would avoid that panic at the end. What do you think?

Reply:
This suggestion seems a bit contradictory to me :P

A v4 loader program account is not the same as a v2 program account.
So, if we compare `loader_id` only with `bpf_loader_upgradeable::ID` we are not going to support v4 programs.
We would just do the wrong thing if anyone sets the `loader_id` to anything but `bpf_loader_upgradeable::ID` or `bpf_loader::ID`.
I do not see how this makes things simpler %)

And just in general, I think, the code is easier to read when cases are explicit.
Rather than when different branches are merged together, and nested `if` statements are used for the distinct parts.
In this particular case, as we now return different enum branches, the code would get pretty hard to follow if we try to combine different cases.

Actually, it might be better to use  `V2`, and `V3` (and, later, `V4`) as the branch names in the `ProgramAccounts` enum, rather than `Upgradeable`/`NonUpgradeable`.
I have renamed the branches.

As for the `panic` at the end - I think it is a good check in any case.
As the `loader_id` value could be anything, it would be wrong to continue execution if the program loader is something this function was not written to support.
So, I think, it is better to add the `panic` check.

Reply:
Alright that's fair. It seems you're making the argument that the original code would need updating anyway, since it was written before V2 was deprecated and V4 was written. Makes sense to me.

This is really a test-only crate, so it's not as crucial, but why not consider adding this error to the `ProgramLoadError` and making this function return a `Result`?

Reply:
`Result` should be used when the caller can do something meaningful with the error.
For example, if it is something that the caller did incorrectly.
Or if the caller would want to handle the error.

`spl_program` returns a `Result`:

```rust
#[derive(Error, Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProgramLoadError {
    #[error("Unknown program address: {0}")]
    UnknownAddress(Pubkey),
}

pub fn spl_program(
    rent: &Rent,
    target_address: Pubkey,
) -> Result<(ProgramElf, ProgramAccounts), ProgramLoadError> {
```

It is used to indicate to the caller that they are trying to load an unknown program.

But an invalid loader is something internal to this module.
The caller can not do anything about it and is not expected to know that this can even happen.
This is a convention between the definition of

```rust
static SPL_PROGRAMS: &[ProgramInfo] = &[ ... ];
```

and `program_accounts` being able to construct accounts for all the loaders specified in the `SPL_PROGRAMS` array.

So, I do not think that returning an error to the caller in case an unexpected loader was specified is correct here.
There is nothing the caller could do about it.
It would only force all caller to write extra error handing code they do not care about.